### ISSUE TITLE
Enable strict CSP headers to sandbox against any possible XSS

### ISSUE THREAD

**Issue by cjdelisle** [vector-im/riot-web#3632]
Modern browsers allow sending Content-Security-Policy headers and will prevent eval() or inline `<script>` tags as well as all of the variants of onload, onerror etc. which is a formidable barrier to XSS security breach.

This is an improvement suggestion which consists of removing any inline `<script>`, eval() or on* javascript (move these to remote loaded `<script>` and event handlers). Then begin sending CSP headers which (for script) allow 'self' and the urls of any content delivery networks which are used.

The benefit is peace of mind because modern browsers will disallow execution of XSS attack script so even if a vulnerability is discovered, the result the the vast majority of users is just an error in the console. Support: http://caniuse.com/#feat=contentsecuritypolicy

**Comment by lampholder** [comment]
This is P3 because it's not on our immediate roadmap, but it seems like a solid idea.

I don't have a real idea for the level of effort or return on investment for Riot Web, though - I'll poke some people for some input.

**Comment by btittelbach** [comment]
You've probably all read the articles where people are now successfully looking into finding remote code execution bugs in Electron apps. https://statuscode.ch/2017/11/from-markdown-to-rce-in-atom/

Would be good to know riot.im is prepared for that.

**Comment by ara4n** [comment]
We use CSP already in the media repository to stop content uploaded stuff containing XSS.  Riot/Web itself isn't served with a CSP however, and unsure how Electron is configured.  Bumping priority accordingly to investigate.

**Comment by ara4n** [comment]
marking as critical until proven otherwise

**Comment by rugk** [comment]
BTW to make sure each user of Riot uses it, I highly suggest to (also) use the in-HTML version of your CSP. (It's just a meta tag, so that is a nice enhancement)

I'll look into creating a CSP, which works with the current version. As always be aware, that [CSPs](content-security-policy.com) are just _additional_ security features, they can mitigate the attack, but they do not replace a real fix of an XSS attack or so, should one be found.

**Comment by rugk** [comment]
BTW it is not critical, but it is a very good (and highly suggested!) security enhancement.

**Comment by rugk** [comment]
E.g. a good step would be to do not use inline JS (i.e. don't include JS in HTML code). I think Riot works quite good with that, it just shows two errors in the console, but these seem to work:
``` 
Content Security Policy: Die Einstellungen der Seite haben das Laden einer Ressource auf self blockiert ("script-src 'self' 'unsafe-eval'"). Source: (function (ERROR) { const V8_STACK_....
```

And, of course, the `window.vector_indexeddb_worker_script = 'bundles/070e2827d7275f44a591/indexeddb-worker.js';` which is included in script tags in the index.html.

**Comment by rugk** [comment]
Same with inline CSS. If you disable that it just breaks the whole styling.

The good news is, it works with a relatively strong JS prevention (`script-src 'self'`)
The bad news, is, it uses `object` so I have to enable `object-src`. (https://github.com/vector-im/riot-web/issues/6165)

BTW a CSP checker can be found at https://csp-evaluator.withgoogle.com/.

**Comment by rugk** [comment]
Here is a suggestion for a CSP, which works for the current version:
```
Content-Security-Policy: "default-src 'none'; style-src 'self' 'unsafe-inline'; script-src 'self'; img-src 'self' data:; connect-src *; font-src 'self'; media-src 'self'; child-src https://scalar.vector.im usercontent.riot.im blob:; form-action 'self'; object-src 'self'" always;
```

It only allows `https://scalar.vector.im` as the integrations server, but that can be adjusted. To limit the connections to one Matrix/Identity server, you would have to adjust `connect-src`. `object-src` is allowed because of https://github.com/vector-im/riot-web/issues/6165-

**Comment by cjdelisle** [comment]
IMO applying this rule goes a long way toward preventing XSS.

**Comment by gerg5c42g542g2c54g52c** [comment]
'unsafe-eval' seems to be needed too, without it CSP reports: call to eval() or related function blocked by CSP. 1 bundle.js:80

Another error CSP reports is this:
`Content Security Policy: The page’s settings blocked the loading of a resource at self (“script-src host 'unsafe-eval'”). Source: 
                window.vector_indexeddb.... 1 host:46`

Any idea how to fix this error?

This is the header with unsafe-eval enabled and only self hosted Identity server enabled that I use:
`Content-Security-Policy: "default-src 'none'; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-eval'; img-src 'self' data:; connect-src 'self'; font-src 'self'; media-src 'self'; child-src https://usercontent.riot.im blob:; form-action 'self'; object-src 'self'"`

**Comment by rugk** [comment]
However, that block does not result in problems, as far as I see. So maybe only something unimportant is blocked.

**Comment by gerg5c42g542g2c54g52c** [comment]
I think it would be a good idea not to block parts of riot-web's code. Plus, I wonder why it's blocked in the first place. Scripts from that host should be allowed.

**Comment by t3chguy** [comment]
I'm not sure that can be moved to js trivially because webpack. As I understand it that line is important so that indexeddb operations happen in a worker. Indexeddb is needed for things like e2e.

**Comment by rugk** [comment]
No, e2e and such stuff worked for me without that line. Maybe it saved the database somehow else/used some fallback, or so…

**Comment by t3chguy** [comment]
Post the console of the app starting up, it'll say if it used a fallback storage

**Comment by rugk** [comment]
Did never look at that exactly, but that also does not matter.

**Comment by t3chguy** [comment]
If its using fallback storage then it does matter as the fallback has race conditions due to being synchronous and localStorage backed which causes e2e to fail in certain circumstances

**Comment by rugk** [comment]
So finally the JS should just be moved to a single file or so. Can't that hard… it's one line, after all.

**Comment by t3chguy** [comment]
Except half of that line *isn't* JavaScript and is instead a build tool generator placeholder for webpack.
`<%= htmlWebpackPlugin.files.js[i] %>`

**Comment by rugk** [comment]
Ah, that's bad then.

**Comment by indolering** [comment]
> I'll look into creating a CSP, which works with the current version. 

Apply the CSP that you have come up with to prevent further technical debt and break this ticket up into further refinements.

> As always be aware, that CSPs are just additional security features, they can mitigate the attack, but they do not replace a real fix of an XSS attack or so, should one be found.

AFAIK CSP is supposed to be the fundamental fix for XSS attacks.  If you lockdown all of your scripts using CSP, then an attacker has to find a browser bug to get around the CSP filter.  I was pretty ... dismissive of Riot's commitment to security when I found out it was relying on

**Comment by indolering** [comment]
>  Except half of that line isn't JavaScript and is instead a build tool generator placeholder for webpack.
<%= htmlWebpackPlugin.files.js[i] %>
and as you can see, its part of the htmlWebpackPlugin which suggests it'll only work in a html file

Just add a build step to push all inline JavaScript to external files.  WebPack apparently [supports CSP nonces](https://webpack.js.org/guides/csp/) natively.

**Comment by rugk** [comment]
> AFAIK CSP is supposed to be the fundamental fix for XSS attacks. If you lockdown all of your scripts using CSP, then an attacker has to find a browser bug to get around the CSP filter.

E.g. there are still older browsers that do not support CSP. Additionally, when an XSS exploit is found, maybe an attacker can also find a way to circumvent your CSP. You just can never be 100% sure that your CSP mitigates (note I do not say "prevents" here deliberately) all XSS attacks. Especially as you cannot disallow everything and you just need to allow something for your application to work. Depending on your application this can be more or (better) less, but it may never be 100% "XSS-proof". At least, you cannot easily judge it. That said, one thing is for sure: A good CSP likely prevents many XSS attacks.

**Comment by cjdelisle** [comment]
@rugk I disagree, IMO a good CSP policy is a water-tight solution to XSS attacks, except on old browsers which probably have known vulnerabilities anyway. Furthermore, a good CSP policy makes XSS attacks much less interesting because they only target the 1% of users who refuse to upgrade their Windows XP machine, so attackers may never bother weaponizing them. Finally, a CSP policy can be configured to report violations to a server, so any attempted attack will be detected as soon as it is seen by a modern browser. Unsanitized html injection should still be considered a bug, but with CSP it is absolutely not something to panic over.

**Comment by rugk** [comment]
All you say is true, sure. Except that it is no "water-tight solution", because of one simple thing:
> IMO a good CSP policy is a water-tight solution to XSS attacks,

How do you define "good"? What is the threshold at which a policy is not good anymore?

And yes, you can say you mean these with `unsafe`, but that may not be enough.
Because, sure, preventing simple HTML injection is easy with CSPs if done properly, but there are many more ways to theoretically and - if a clever attacker/security researcher discovers it - practically exploit it. See [this Stackexchange answer](https://security.stackexchange.com/questions/86561/is-csp-enough-to-prevent-xss-attacks#86564) and [this here](https://security.stackexchange.com/questions/38001/xss-prevention-through-content-security-policy) for details.

So yes, you may not need to panic - that does not help anyway, in most cases -, but XSS is a big topic. As such, you still need to treat any XSS vulnerability as a vulnerability - because it is one. A CSP is one layer of defense, in this case likely the last one, but that must not prevent you from fixing the underlying vulnerabilities, even if they are migrated by a CSP.

**Comment by indolering** [comment]
This has been a fun discussion on the state of XSS prevention, but can we all agree that CSP is *super* important and move on to implementing this?

@rugk, can you fix your CSP policy by adding `unsafe-eval` and get it mainlined?  Then we can start new tickets targeting each bad policy and prevent further security regressions.

**Comment by rugk** [comment]
In my tests everything worked without `unsafe-eval`. Anyway, it's of course only one example. Feel free to add whatever you want and remove parts to get stricter and stricter by time.

**Comment by indolering** [comment]
> In my tests everything worked without unsafe-eval. Anyway, it's of course only one example. Feel free to add whatever you want and remove parts to get stricter and stricter by time.

It sounds like it won't be mainlined without that piece of the codebase working.  Could you add a build step that moves the inline script to an external file?  Again, the important piece here is to get a working CSP into the codebase so we can improve things incrementally, let's not make this a blocking issue.

**Comment by rugk** [comment]
Ah, so the Riot devs should add it, sure… [as you've mentioned how this can be done before](https://github.com/vector-im/riot-web/issues/3632#issuecomment-402298436).

**Comment by indolering** [comment]
> Ah, so the Riot devs should add it, sure… as you've mentioned how this can be done before.

I'm basically asking if you will make the pull request or do I have to?

**Comment by rugk** [comment]
No, I'll not do a PR.

**Comment by Nadrieril** [comment]
Is there a recommended CSP string I can safely use when selfhosting riot ? And should this be documented somewhere ?

**Comment by rugk** [comment]
@Nadrieril I've added mine/one [in here](https://github.com/vector-im/riot-web/issues/3632#issuecomment-367110054).

It should preferably not be documented, but actually tweaked (if needed) and merged into master.

**Comment by Nadrieril** [comment]
Ok, thanks ! That'd be very cool indeed

**Comment by indolering** [comment]
> It should preferably not be documented, but actually tweaked (if needed) and merged into the code.

I would document and unit test each rule* to make sure it hasn't been borked!  Declarative specs count as code 😄.

*This is technically unit testing and might require a service like Sauce Labs (which has free accounts for F/OSS projects).

**Comment by indolering** [comment]
@rugk Did you ever get around to that PR?

**Comment by rugk** [comment]
As explained earlier [no](https://github.com/vector-im/riot-web/issues/3632#issuecomment-403928388). Especially, because I don't know it is what the riot devs want. (and how strict it should be, testing etc.)

**Comment by 532910** [comment]
https://observatory.mozilla.org/analyze/riot.im
Grade F

**Comment by ara4n** [comment]
riot.im/app now has a CSP on it. however, we have `script-src: unsafe-inline` due to the webpack issue mentioned [above](https://github.com/vector-im/riot-web/issues/3632#issuecomment-381179249) - need to fix this before we can close it up (and add it to riot-web as an meta tag)

**Comment by aaronraimist** [comment]
For the record the current CSP is, as of ^ is

```
default-src 'none';
style-src 'self' 'unsafe-inline';
script-src 'self' 'unsafe-eval' 'unsafe-inline' https://www.recaptcha.net https://www.gstatic.com;
img-src * blob: data:;
connect-src *;
font-src 'self';
media-src * blob: data:;
child-src * blob: data:;
form-action 'self';
object-src 'self';
manifest-src 'self'
```

**Comment by ara4n** [comment]
...and we need to add the CSP as a meta tag or similar so that Electron picks it up.

**Comment by aaronraimist** [comment]
Related: #11610

**Comment by stoically** [comment]
Had a quick glance why `unsafe-eval` currently (v1.5.8) is required and found 18 instances in the final bundle. It seems there are 13 instances of `new Function` which function as fallback if the environment isn't an actual browser, so for those dropping `unsafe-eval` would probably be fine. Then there are 5 instances of `eval` which come from [Modernizer trying to detect es5](https://github.com/Modernizr/Modernizr/blob/25f2f3d0030a828f03b533f47a61da514e2e5139/feature-detects/es5/syntax.js#L13), not sure what's the impact of that not working.

I'm using riot web without `unsafe-eval` for a while now and haven't noticed any problems.

**Comment by jryans** [comment]
In addition to embedding the CSP as a meta tag in the app, we'd also like to remove `unsafe-inline` from `script-src`.

**Comment by t3chguy** [comment]
Splitting off a new issue to track getting rid of `unsafe-eval` which is currently required for WASM.
